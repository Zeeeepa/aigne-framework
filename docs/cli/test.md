# Testing

Use AIGNE CLI to run agent tests.

## `aigne test` Command

Run tests in the specified agent directory.

```bash
aigne test [options]
```

### Options

* `--path <path>`: Path to the agent directory (defaults to current directory `.`)
* `--help`: Show command help

### Examples

#### Run tests in current directory

```bash
aigne test
```

#### Run tests in specific directory

```bash
aigne test --path ./my-agents
```

### Test File Structure

The AIGNE testing system automatically finds and runs test files in your project. Make sure your test files follow the correct naming conventions and structure.

### Best Practices

1. **Organize test files**: Place test files in appropriate directory structures
2. **Write clear test cases**: Ensure each test case has a clear purpose and expected outcome
3. **Use descriptive test names**: Make test names clearly express what the test covers
4. **Run tests regularly**: Run tests regularly during development to ensure code quality
