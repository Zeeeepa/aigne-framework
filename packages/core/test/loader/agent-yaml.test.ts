import { expect, spyOn, test } from "bun:test";
import assert from "node:assert";
import { join } from "node:path";
import {
  type Agent,
  AIAgent,
  AIGNE,
  ProcessMode,
  stringToAgentResponseStream,
  TeamAgent,
  TransformAgent,
} from "@aigne/core";
import { loadAgentFromYamlFile } from "@aigne/core/loader/agent-yaml.js";
import { loadAgent } from "@aigne/core/loader/index.js";
import { pick } from "@aigne/core/utils/type-utils.js";
import { nodejs } from "@aigne/platform-helpers/nodejs/index.js";
import { ZodType } from "zod";
import zodToJsonSchema from "zod-to-json-schema";
import { OpenAIChatModel } from "../_mocks/mock-models.js";

test("loadAgentFromYaml should load AIAgent correctly", async () => {
  const agent = await loadAgent(join(import.meta.dirname, "../../test-agents/chat.yaml"));

  expect(agent).toBeInstanceOf(AIAgent);
  assert(agent instanceof AIAgent, "agent should be an instance of AIAgent");

  expect({
    name: agent.name,
    alias: agent.alias,
    description: agent.description,
    instructions: agent.instructions.instructions,
    skills: agent.skills.map((skill) => ({
      name: skill.name,
      description: skill.description,
      input_schema: zodToJsonSchema(skill.inputSchema),
      output_schema: zodToJsonSchema(skill.outputSchema),
      default_input: skill.defaultInput,
    })),
  }).toMatchSnapshot();
});

test("loadAgentFromYaml should error if agent.yaml file is invalid", async () => {
  spyOn(nodejs.fs, "readFile")
    .mockReturnValueOnce(Promise.reject(new Error("no such file or directory")))
    .mockReturnValueOnce(Promise.resolve("[this is not a valid yaml}"))
    .mockReturnValueOnce(Promise.resolve("name: 123"));

  expect(loadAgentFromYamlFile("./not-exist-aigne.yaml")).rejects.toThrow(
    "no such file or directory",
  );

  expect(loadAgentFromYamlFile("./invalid-aigne.yaml")).rejects.toThrow(
    "Failed to parse agent definition",
  );

  expect(loadAgentFromYamlFile("./invalid-content-aigne.yaml")).rejects.toThrow(
    "Failed to validate agent definition",
  );
});

test("loadAgentFromYaml should load mcp agent correctly", async () => {
  spyOn(nodejs.fs, "readFile")
    .mockReturnValueOnce(
      Promise.resolve(`\
type: mcp
url: http://localhost:3000/sse
`),
    )
    .mockReturnValueOnce(
      Promise.resolve(`\
type: mcp
command: npx
args: ["-y", "@modelcontextprotocol/server-filesystem", "."]
`),
    );

  expect(await loadAgentFromYamlFile("./remote-mcp.yaml")).toEqual({
    type: "mcp",
    url: "http://localhost:3000/sse",
  });

  expect(await loadAgentFromYamlFile("./local-mcp.yaml")).toEqual({
    type: "mcp",
    command: "npx",
    args: ["-y", "@modelcontextprotocol/server-filesystem", "."],
  });
});

test("loadAgentFromYaml should load TeamAgent correctly", async () => {
  const agent = await loadAgent(join(import.meta.dirname, "../../test-agents/team.yaml"));

  expect(agent).toBeInstanceOf(TeamAgent);
  assert(agent instanceof TeamAgent, "agent should be an instance of AIAgent");

  expect(agent).toEqual(
    expect.objectContaining({
      name: "test-team-agent",
      description: "Test team agent",
      mode: ProcessMode.parallel,
    }),
  );

  expect(agent.skills.length).toBe(2);
  expect(agent.iterateOn).toBe("sections");
});

test("loadAgentFromYaml should load AIAgent with prompt file correctly", async () => {
  const model = new OpenAIChatModel();

  const aigne = new AIGNE({ model });

  const agent = await loadAgent(
    join(import.meta.dirname, "../../test-agents/chat-with-prompt.yaml"),
  );

  expect(agent).toBeInstanceOf(AIAgent);
  assert(agent instanceof AIAgent, "agent should be an instance of AIAgent");

  expect(agent.instructions.instructions).toMatchSnapshot();

  const modelProcess = spyOn(model, "process").mockReturnValueOnce(
    stringToAgentResponseStream("Hello, this is a test response message"),
  );
  await aigne.invoke(agent, {
    language: "English",
  });
  expect(pick(modelProcess.mock.lastCall?.at(0) as any, "messages")).toMatchSnapshot({});
});

test("loadAgentFromYaml should load nested agent correctly", async () => {
  const agent = await loadAgent(join(import.meta.dirname, "../../test-agents/nested-agent.yaml"));

  expect(agent).toBeInstanceOf(TeamAgent);
  expect(agent.name).toMatchInlineSnapshot(`"test-nested-agent"`);
  expect(agent.description).toMatchInlineSnapshot(`"Test nested agent"`);

  const flatten = (agent: Agent): unknown => {
    return {
      name: agent.name,
      instructions: agent instanceof AIAgent ? agent.instructions?.instructions : undefined,
      inputSchema:
        agent["_inputSchema"] instanceof ZodType
          ? zodToJsonSchema(agent["_inputSchema"])
          : undefined,
      outputSchema:
        agent["_outputSchema"] instanceof ZodType
          ? zodToJsonSchema(agent["_outputSchema"])
          : undefined,
      skills: agent.skills.map((i) => flatten(i)),
    };
  };

  expect(flatten(agent)).toMatchSnapshot();
});

test("loadAgentFromYaml should load transform agent correctly", async () => {
  const agent = await loadAgent(join(import.meta.dirname, "../../test-agents/transform.yaml"));

  expect(agent).toBeInstanceOf(TransformAgent);
  assert(agent instanceof TransformAgent);
  expect(agent.name).toMatchInlineSnapshot(`"transform-agent"`);
  expect(agent.description).toMatchInlineSnapshot(`
    "A Transform Agent that processes input data using JSONata expressions.
    "
  `);
  expect(agent["jsonata"]).toMatchInlineSnapshot(`
    "{
      userId: user_id,
      userName: user_name,
      createdAt: created_at
    }
    "
  `);
});

test("loadAgentFromYaml should load external schema agent correctly", async () => {
  const agent = await loadAgent(
    join(import.meta.dirname, "../../test-agents/external-schema-agent.yaml"),
  );

  expect(zodToJsonSchema(agent.inputSchema)).toMatchSnapshot();
  expect(zodToJsonSchema(agent.outputSchema)).toMatchSnapshot();
});

test("loadAgentFromYaml should load nested external schema agent correctly", async () => {
  const agent = await loadAgent(
    join(import.meta.dirname, "../../test-agents/external-schema-agent-nested.yaml"),
  );

  expect(zodToJsonSchema(agent.inputSchema)).toMatchSnapshot();
  expect(zodToJsonSchema(agent.outputSchema)).toMatchSnapshot();
});

test("loadAgentFromYaml should support various styles of naming", async () => {
  const agent = await loadAgent(
    join(import.meta.dirname, "../../test-agents/test-agent-input-naming.yaml"),
  );

  expect(zodToJsonSchema(agent.inputSchema)).toMatchSnapshot();
  expect(zodToJsonSchema(agent.outputSchema)).toMatchSnapshot();
});

test("loadAgentFromYaml should support default input", async () => {
  const agent = await loadAgent(
    join(import.meta.dirname, "../../test-agents/test-agent-with-default-input.yaml"),
  );

  expect({
    name: agent.name,
    skills: agent.skills.map((i) => ({
      name: i.name,
      input_schema: zodToJsonSchema(i.inputSchema),
      default_input: i.defaultInput,
    })),
  }).toMatchSnapshot();
});

test("loadAgentFromYaml should support hooks", async () => {
  const agent = await loadAgent(
    join(import.meta.dirname, "../../test-agents/test-agent-with-hooks.yaml"),
  );

  expect({
    name: agent.name,
    hooks: agent.hooks.map((i) =>
      Object.fromEntries(Object.entries(i).map(([k, v]) => [k, k === "priority" ? v : v?.name])),
    ),
    skills: agent.skills.map((i) => ({
      name: i.name,
      hooks: agent.hooks.map((i) =>
        Object.fromEntries(Object.entries(i).map(([k, v]) => [k, k === "priority" ? v : v?.name])),
      ),
    })),
  }).toMatchSnapshot();
});

test("loadAgentFromYaml should support reflection for TeamAgent", async () => {
  const agent = await loadAgent(
    join(import.meta.dirname, "../../test-agents/team-agent-with-reflection.yaml"),
  );

  assert(agent instanceof TeamAgent, "agent should be an instance of TeamAgent");

  expect({
    name: agent.name,
    skills: agent.skills.map((i) => ({
      name: i.name,
    })),
    reflection: agent.reflection && {
      ...agent.reflection,
      reviewer: {
        name: agent.reflection.reviewer.name,
      },
    },
  }).toMatchSnapshot();
});

test("loadAgentFromYaml should support reflection for TeamAgent with inline reviewer", async () => {
  const agent = await loadAgent(
    join(import.meta.dirname, "../../test-agents/team-agent-with-reflection-inline.yaml"),
  );

  assert(agent instanceof TeamAgent, "agent should be an instance of TeamAgent");

  expect({
    name: agent.name,
    skills: agent.skills.map((i) => ({
      name: i.name,
    })),
    reflection: agent.reflection && {
      ...agent.reflection,
      reviewer: {
        name: agent.reflection.reviewer.name,
      },
    },
  }).toMatchSnapshot();
});
