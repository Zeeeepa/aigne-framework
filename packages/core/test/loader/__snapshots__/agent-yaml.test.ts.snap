// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`loadAgentFromYaml should load AIAgent correctly 1`] = `
{
  "alias": [
    "chat-bot",
    "bot",
  ],
  "description": "Chat agent",
  "instructions": 
"You are a helpful assistant that can answer questions and provide information on a wide range of topics.
Your goal is to assist users in finding the information they need and to engage in friendly conversation.
"
,
  "name": "chat",
  "skills": [
    {
      "default_input": undefined,
      "description": "This agent evaluates JavaScript code.",
      "input_schema": {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "additionalProperties": true,
        "properties": {
          "code": {
            "description": "JavaScript code to evaluate",
            "type": "string",
          },
        },
        "required": [
          "code",
        ],
        "type": "object",
      },
      "name": "evaluateJs",
      "output_schema": {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "additionalProperties": true,
        "properties": {
          "result": {
            "description": "Result of the evaluated code",
          },
        },
        "type": "object",
      },
    },
  ],
}
`;

exports[`loadAgentFromYaml should load AIAgent with prompt file correctly 1`] = `
"You are a helper agent to answer everything about chat prompt in AIGNE.

{% include "language_instruction.txt" %}
"
`;

exports[`loadAgentFromYaml should load AIAgent with prompt file correctly 2`] = `
{
  "messages": [
    {
      "content": 
"You are a helper agent to answer everything about chat prompt in AIGNE.

Output in English

"
,
      "name": undefined,
      "role": "system",
    },
  ],
}
`;

exports[`loadAgentFromYaml should load nested agent correctly 3`] = `
{
  "inputSchema": undefined,
  "instructions": undefined,
  "name": "test-nested-agent",
  "outputSchema": undefined,
  "skills": [
    {
      "inputSchema": undefined,
      "instructions": 
"You are a helpful assistant that can answer questions and provide information on a wide range of topics.
Your goal is to assist users in finding the information they need and to engage in friendly conversation.
"
,
      "name": "chat",
      "outputSchema": undefined,
      "skills": [
        {
          "inputSchema": {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "additionalProperties": false,
            "properties": {
              "code": {
                "description": "JavaScript code to evaluate",
                "type": "string",
              },
            },
            "required": [
              "code",
            ],
            "type": "object",
          },
          "instructions": undefined,
          "name": "evaluateJs",
          "outputSchema": {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "additionalProperties": false,
            "properties": {
              "result": {
                "description": "Result of the evaluated code",
              },
            },
            "type": "object",
          },
          "skills": [],
        },
      ],
    },
    {
      "inputSchema": undefined,
      "instructions": 
"You are a helper agent to answer everything about chat prompt in AIGNE.

{% include "language_instruction.txt" %}
"
,
      "name": "child",
      "outputSchema": undefined,
      "skills": [],
    },
    {
      "inputSchema": undefined,
      "instructions": undefined,
      "name": "nested-team",
      "outputSchema": undefined,
      "skills": [
        {
          "inputSchema": undefined,
          "instructions": 
"You are a helpful assistant that can answer questions and provide information on a wide range of topics.
Your goal is to assist users in finding the information they need and to engage in friendly conversation.
"
,
          "name": "chat",
          "outputSchema": undefined,
          "skills": [
            {
              "inputSchema": {
                "$schema": "http://json-schema.org/draft-07/schema#",
                "additionalProperties": false,
                "properties": {
                  "code": {
                    "description": "JavaScript code to evaluate",
                    "type": "string",
                  },
                },
                "required": [
                  "code",
                ],
                "type": "object",
              },
              "instructions": undefined,
              "name": "evaluateJs",
              "outputSchema": {
                "$schema": "http://json-schema.org/draft-07/schema#",
                "additionalProperties": false,
                "properties": {
                  "result": {
                    "description": "Result of the evaluated code",
                  },
                },
                "type": "object",
              },
              "skills": [],
            },
          ],
        },
        {
          "inputSchema": {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "additionalProperties": false,
            "properties": {
              "message": {
                "description": "Message to process",
                "type": "string",
              },
            },
            "required": [
              "message",
            ],
            "type": "object",
          },
          "instructions": 
"You are a helper agent to answer everything about chat prompt in AIGNE.

{% include "language_instruction.txt" %}
"
,
          "name": "nested-child",
          "outputSchema": {
            "$schema": "http://json-schema.org/draft-07/schema#",
            "additionalProperties": false,
            "properties": {
              "response": {
                "description": "Response from nested child agent",
                "type": "string",
              },
            },
            "required": [
              "response",
            ],
            "type": "object",
          },
          "skills": [],
        },
      ],
    },
  ],
}
`;

exports[`loadAgentFromYaml should load external schema agent correctly 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": true,
  "properties": {
    "first": {
      "description": "First level property",
      "type": "string",
    },
    "second": {
      "items": {
        "additionalProperties": false,
        "properties": {
          "array": {
            "items": {
              "additionalProperties": false,
              "properties": {
                "third": {
                  "description": "Third level property",
                  "type": "string",
                },
              },
              "required": [
                "third",
              ],
              "type": "object",
            },
            "type": "array",
          },
          "second": {
            "description": "Second level property",
            "type": "string",
          },
          "third": {
            "additionalProperties": false,
            "properties": {
              "third": {
                "description": "Third level property",
                "type": "string",
              },
            },
            "required": [
              "third",
            ],
            "type": "object",
          },
        },
        "required": [
          "second",
          "third",
          "array",
        ],
        "type": "object",
      },
      "type": "array",
    },
  },
  "required": [
    "first",
    "second",
  ],
  "type": "object",
}
`;

exports[`loadAgentFromYaml should load external schema agent correctly 2`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": true,
  "properties": {
    "first": {
      "description": "First level property",
      "type": "string",
    },
    "second": {
      "items": {
        "additionalProperties": false,
        "properties": {
          "array": {
            "items": {
              "additionalProperties": false,
              "properties": {
                "third": {
                  "description": "Third level property",
                  "type": "string",
                },
              },
              "required": [
                "third",
              ],
              "type": "object",
            },
            "type": "array",
          },
          "second": {
            "description": "Second level property",
            "type": "string",
          },
          "third": {
            "additionalProperties": false,
            "properties": {
              "third": {
                "description": "Third level property",
                "type": "string",
              },
            },
            "required": [
              "third",
            ],
            "type": "object",
          },
        },
        "required": [
          "second",
          "third",
          "array",
        ],
        "type": "object",
      },
      "type": "array",
    },
  },
  "required": [
    "first",
    "second",
  ],
  "type": "object",
}
`;

exports[`loadAgentFromYaml should load nested external schema agent correctly 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": true,
  "properties": {
    "array": {
      "items": {
        "additionalProperties": false,
        "properties": {
          "external": {
            "additionalProperties": false,
            "properties": {
              "first": {
                "description": "First level property",
                "type": "string",
              },
              "second": {
                "items": {
                  "additionalProperties": false,
                  "properties": {
                    "array": {
                      "items": {
                        "additionalProperties": false,
                        "properties": {
                          "third": {
                            "description": "Third level property",
                            "type": "string",
                          },
                        },
                        "required": [
                          "third",
                        ],
                        "type": "object",
                      },
                      "type": "array",
                    },
                    "second": {
                      "description": "Second level property",
                      "type": "string",
                    },
                    "third": {
                      "additionalProperties": false,
                      "properties": {
                        "third": {
                          "description": "Third level property",
                          "type": "string",
                        },
                      },
                      "required": [
                        "third",
                      ],
                      "type": "object",
                    },
                  },
                  "required": [
                    "second",
                    "third",
                    "array",
                  ],
                  "type": "object",
                },
                "type": "array",
              },
            },
            "required": [
              "first",
              "second",
            ],
            "type": "object",
          },
        },
        "required": [
          "external",
        ],
        "type": "object",
      },
      "type": "array",
    },
    "external": {
      "additionalProperties": false,
      "properties": {
        "first": {
          "description": "First level property",
          "type": "string",
        },
        "second": {
          "items": {
            "additionalProperties": false,
            "properties": {
              "array": {
                "items": {
                  "additionalProperties": false,
                  "properties": {
                    "third": {
                      "description": "Third level property",
                      "type": "string",
                    },
                  },
                  "required": [
                    "third",
                  ],
                  "type": "object",
                },
                "type": "array",
              },
              "second": {
                "description": "Second level property",
                "type": "string",
              },
              "third": {
                "additionalProperties": false,
                "properties": {
                  "third": {
                    "description": "Third level property",
                    "type": "string",
                  },
                },
                "required": [
                  "third",
                ],
                "type": "object",
              },
            },
            "required": [
              "second",
              "third",
              "array",
            ],
            "type": "object",
          },
          "type": "array",
        },
      },
      "required": [
        "first",
        "second",
      ],
      "type": "object",
    },
    "items": {
      "items": {
        "additionalProperties": false,
        "properties": {
          "first": {
            "description": "First level property",
            "type": "string",
          },
          "second": {
            "items": {
              "additionalProperties": false,
              "properties": {
                "array": {
                  "items": {
                    "additionalProperties": false,
                    "properties": {
                      "third": {
                        "description": "Third level property",
                        "type": "string",
                      },
                    },
                    "required": [
                      "third",
                    ],
                    "type": "object",
                  },
                  "type": "array",
                },
                "second": {
                  "description": "Second level property",
                  "type": "string",
                },
                "third": {
                  "additionalProperties": false,
                  "properties": {
                    "third": {
                      "description": "Third level property",
                      "type": "string",
                    },
                  },
                  "required": [
                    "third",
                  ],
                  "type": "object",
                },
              },
              "required": [
                "second",
                "third",
                "array",
              ],
              "type": "object",
            },
            "type": "array",
          },
        },
        "required": [
          "first",
          "second",
        ],
        "type": "object",
      },
      "type": "array",
    },
    "nested": {
      "additionalProperties": false,
      "properties": {
        "external": {
          "additionalProperties": false,
          "properties": {
            "first": {
              "description": "First level property",
              "type": "string",
            },
            "second": {
              "items": {
                "additionalProperties": false,
                "properties": {
                  "array": {
                    "items": {
                      "additionalProperties": false,
                      "properties": {
                        "third": {
                          "description": "Third level property",
                          "type": "string",
                        },
                      },
                      "required": [
                        "third",
                      ],
                      "type": "object",
                    },
                    "type": "array",
                  },
                  "second": {
                    "description": "Second level property",
                    "type": "string",
                  },
                  "third": {
                    "additionalProperties": false,
                    "properties": {
                      "third": {
                        "description": "Third level property",
                        "type": "string",
                      },
                    },
                    "required": [
                      "third",
                    ],
                    "type": "object",
                  },
                },
                "required": [
                  "second",
                  "third",
                  "array",
                ],
                "type": "object",
              },
              "type": "array",
            },
          },
          "required": [
            "first",
            "second",
          ],
          "type": "object",
        },
      },
      "required": [
        "external",
      ],
      "type": "object",
    },
  },
  "required": [
    "external",
    "nested",
    "items",
    "array",
  ],
  "type": "object",
}
`;

exports[`loadAgentFromYaml should load nested external schema agent correctly 2`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": true,
  "properties": {
    "array": {
      "items": {
        "additionalProperties": false,
        "properties": {
          "external": {
            "additionalProperties": false,
            "properties": {
              "first": {
                "description": "First level property",
                "type": "string",
              },
              "second": {
                "items": {
                  "additionalProperties": false,
                  "properties": {
                    "array": {
                      "items": {
                        "additionalProperties": false,
                        "properties": {
                          "third": {
                            "description": "Third level property",
                            "type": "string",
                          },
                        },
                        "required": [
                          "third",
                        ],
                        "type": "object",
                      },
                      "type": "array",
                    },
                    "second": {
                      "description": "Second level property",
                      "type": "string",
                    },
                    "third": {
                      "additionalProperties": false,
                      "properties": {
                        "third": {
                          "description": "Third level property",
                          "type": "string",
                        },
                      },
                      "required": [
                        "third",
                      ],
                      "type": "object",
                    },
                  },
                  "required": [
                    "second",
                    "third",
                    "array",
                  ],
                  "type": "object",
                },
                "type": "array",
              },
            },
            "required": [
              "first",
              "second",
            ],
            "type": "object",
          },
        },
        "required": [
          "external",
        ],
        "type": "object",
      },
      "type": "array",
    },
    "external": {
      "additionalProperties": false,
      "properties": {
        "first": {
          "description": "First level property",
          "type": "string",
        },
        "second": {
          "items": {
            "additionalProperties": false,
            "properties": {
              "array": {
                "items": {
                  "additionalProperties": false,
                  "properties": {
                    "third": {
                      "description": "Third level property",
                      "type": "string",
                    },
                  },
                  "required": [
                    "third",
                  ],
                  "type": "object",
                },
                "type": "array",
              },
              "second": {
                "description": "Second level property",
                "type": "string",
              },
              "third": {
                "additionalProperties": false,
                "properties": {
                  "third": {
                    "description": "Third level property",
                    "type": "string",
                  },
                },
                "required": [
                  "third",
                ],
                "type": "object",
              },
            },
            "required": [
              "second",
              "third",
              "array",
            ],
            "type": "object",
          },
          "type": "array",
        },
      },
      "required": [
        "first",
        "second",
      ],
      "type": "object",
    },
    "items": {
      "items": {
        "additionalProperties": false,
        "properties": {
          "first": {
            "description": "First level property",
            "type": "string",
          },
          "second": {
            "items": {
              "additionalProperties": false,
              "properties": {
                "array": {
                  "items": {
                    "additionalProperties": false,
                    "properties": {
                      "third": {
                        "description": "Third level property",
                        "type": "string",
                      },
                    },
                    "required": [
                      "third",
                    ],
                    "type": "object",
                  },
                  "type": "array",
                },
                "second": {
                  "description": "Second level property",
                  "type": "string",
                },
                "third": {
                  "additionalProperties": false,
                  "properties": {
                    "third": {
                      "description": "Third level property",
                      "type": "string",
                    },
                  },
                  "required": [
                    "third",
                  ],
                  "type": "object",
                },
              },
              "required": [
                "second",
                "third",
                "array",
              ],
              "type": "object",
            },
            "type": "array",
          },
        },
        "required": [
          "first",
          "second",
        ],
        "type": "object",
      },
      "type": "array",
    },
    "nested": {
      "additionalProperties": false,
      "properties": {
        "external": {
          "additionalProperties": false,
          "properties": {
            "first": {
              "description": "First level property",
              "type": "string",
            },
            "second": {
              "items": {
                "additionalProperties": false,
                "properties": {
                  "array": {
                    "items": {
                      "additionalProperties": false,
                      "properties": {
                        "third": {
                          "description": "Third level property",
                          "type": "string",
                        },
                      },
                      "required": [
                        "third",
                      ],
                      "type": "object",
                    },
                    "type": "array",
                  },
                  "second": {
                    "description": "Second level property",
                    "type": "string",
                  },
                  "third": {
                    "additionalProperties": false,
                    "properties": {
                      "third": {
                        "description": "Third level property",
                        "type": "string",
                      },
                    },
                    "required": [
                      "third",
                    ],
                    "type": "object",
                  },
                },
                "required": [
                  "second",
                  "third",
                  "array",
                ],
                "type": "object",
              },
              "type": "array",
            },
          },
          "required": [
            "first",
            "second",
          ],
          "type": "object",
        },
      },
      "required": [
        "external",
      ],
      "type": "object",
    },
  },
  "required": [
    "external",
    "nested",
    "items",
    "array",
  ],
  "type": "object",
}
`;

exports[`loadAgentFromYaml should support various styles of naming 1`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": true,
  "properties": {
    "PascalCase": {
      "description": "Pascal case property",
      "type": "string",
    },
    "SCREAM_CASE": {
      "description": "Scream case property",
      "type": "string",
    },
    "camelCase": {
      "description": "Camel case property",
      "type": "string",
    },
    "kebab-case": {
      "description": "Kebab case property",
      "type": "string",
    },
    "snake_case": {
      "description": "Snake case property",
      "type": "string",
    },
  },
  "required": [
    "snake_case",
    "camelCase",
    "kebab-case",
    "SCREAM_CASE",
    "PascalCase",
  ],
  "type": "object",
}
`;

exports[`loadAgentFromYaml should support various styles of naming 2`] = `
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": true,
  "properties": {
    "response": {
      "description": "Response from the agent",
      "type": "string",
    },
  },
  "required": [
    "response",
  ],
  "type": "object",
}
`;

exports[`loadAgentFromYaml should support default input 1`] = `
{
  "name": "test_agent_with_default_input",
  "skills": [
    {
      "default_input": {
        "description": {
          "$get": "title",
        },
        "title": "Default Title",
      },
      "input_schema": {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "additionalProperties": true,
        "properties": {
          "description": {
            "description": "Description of the input",
            "type": "string",
          },
          "title": {
            "description": "Title of the input",
            "type": "string",
          },
        },
        "required": [
          "title",
          "description",
        ],
        "type": "object",
      },
      "name": "test_agent_input_naming_skill",
    },
    {
      "default_input": {
        "description": {
          "$get": "title",
        },
        "title": "Default Title JS",
      },
      "input_schema": {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "additionalProperties": true,
        "properties": {
          "description": {
            "type": "string",
          },
          "title": {
            "type": "string",
          },
        },
        "required": [
          "title",
          "description",
        ],
        "type": "object",
      },
      "name": "testAgentWithDefaultInputSkillJS",
    },
    {
      "default_input": {
        "description": {
          "$get": "title",
        },
        "title": "Another Default Title",
      },
      "input_schema": {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "additionalProperties": true,
        "properties": {
          "description": {
            "description": "Description of the input",
            "type": "string",
          },
          "title": {
            "description": "Title of the input",
            "type": "string",
          },
        },
        "required": [
          "title",
          "description",
        ],
        "type": "object",
      },
      "name": "test_agent_with_default_input_skill2.yaml",
    },
  ],
}
`;

exports[`loadAgentFromYaml should support hooks 1`] = `
{
  "hooks": [
    {
      "onEnd": "test_hooks",
      "onError": "test_hooks",
      "onHandoff": "test_hooks",
      "onSkillEnd": "test_hooks",
      "onSkillStart": "test_hooks",
      "onStart": "test_hooks_inline",
      "onSuccess": "test_hooks",
      "priority": "high",
    },
  ],
  "name": "test_agent_with_default_input",
  "skills": [
    {
      "hooks": [
        {
          "onEnd": "test_hooks",
          "onError": "test_hooks",
          "onHandoff": "test_hooks",
          "onSkillEnd": "test_hooks",
          "onSkillStart": "test_hooks",
          "onStart": "test_hooks_inline",
          "onSuccess": "test_hooks",
          "priority": "high",
        },
      ],
      "name": "test_agent_input_naming_skill",
    },
    {
      "hooks": [
        {
          "onEnd": "test_hooks",
          "onError": "test_hooks",
          "onHandoff": "test_hooks",
          "onSkillEnd": "test_hooks",
          "onSkillStart": "test_hooks",
          "onStart": "test_hooks_inline",
          "onSuccess": "test_hooks",
          "priority": "high",
        },
      ],
      "name": "test_agent_with_default_input_skill2.yaml",
    },
  ],
}
`;

exports[`loadAgentFromYaml should support reflection for TeamAgent 1`] = `
{
  "name": "test-team-agent-with-reflection",
  "reflection": {
    "isApproved": "approved",
    "maxIterations": 5,
    "returnLastOnMaxIterations": true,
    "reviewer": {
      "name": "test-team-agent-reviewer",
    },
  },
  "skills": [
    {
      "name": "chat",
    },
  ],
}
`;

exports[`loadAgentFromYaml should support reflection for TeamAgent with inline reviewer 1`] = `
{
  "name": "test-team-agent-with-reflection-inline",
  "reflection": {
    "isApproved": "approved",
    "maxIterations": 5,
    "returnLastOnMaxIterations": true,
    "reviewer": {
      "name": "test-team-agent-reviewer",
    },
  },
  "skills": [
    {
      "name": "chat",
    },
  ],
}
`;
